File: watcher/ingest/normalizer.ts
import type {
  Availability,
  FuelCode,
  FuelState,
  StationSnapshot,
} from "../domain/model.ts";
import { type GeoPoint, haversineKm } from "../domain/geo.ts";
import type { RawStationRecord } from "./raw-record.ts";

const FUEL_FIELDS: Record<FuelCode, string> = {
  Gazole: "gazole",
  SP95: "sp95",
  E10: "e10",
  SP98: "sp98",
  E85: "e85",
  GPLc: "gplc",
};

function numberOrNull(value: unknown): number | null {
  if (typeof value === "number") return Number.isFinite(value) ? value : null;
  if (typeof value === "string" && value.trim() !== "") {
    const parsed = Number(value);
    return Number.isFinite(parsed) ? parsed : null;
  }
  return null;
}

function stringOrNull(value: unknown): string | null {
  if (typeof value !== "string") return null;
  const trimmed = value.trim();
  return trimmed === "" ? null : trimmed;
}

const PARIS_TIME_ZONE = "Europe/Paris";

function zoneOffsetMinutes(utcMs: number): number {
  const parts = new Intl.DateTimeFormat("en-US", {
    timeZone: PARIS_TIME_ZONE,
    hour12: false,
    year: "numeric",
    month: "2-digit",
    day: "2-digit",
    hour: "2-digit",
    minute: "2-digit",
    second: "2-digit",
  }).formatToParts(new Date(utcMs));
  const get = (type: string): number =>
    Number(parts.find((part) => part.type === type)?.value);
  const asUtc = Date.UTC(
    get("year"),
    get("month") - 1,
    get("day"),
    get("hour") % 24,
    get("minute"),
    get("second"),
  );
  return (asUtc - utcMs) / 60000;
}

/**
 * The ODS feed returns French station timestamps as local wall-clock time but
 * labels them UTC (e.g. "2026-09-18T18:49:25+00:00"). Reinterpret the
 * wall-clock as Europe/Paris time so the resulting instant is correct.
 */
export function normalizeFrenchTimestamp(value: string): string | null {
  const match = /^(\d{4})-(\d{2})-(\d{2})[T ](\d{2}):(\d{2})(?::(\d{2}))?/.exec(
    value,
  );
  if (!match) return null;
  const naiveUtc = Date.UTC(
    Number(match[1]),
    Number(match[2]) - 1,
    Number(match[3]),
    Number(match[4]),
    Number(match[5]),
    match[6] ? Number(match[6]) : 0,
  );
  const offset = zoneOffsetMinutes(naiveUtc);
  return new Date(naiveUtc - offset * 60000).toISOString();
}

function timestampOrNull(value: unknown): string | null {
  const raw = stringOrNull(value);
  if (raw === null) return null;
  return normalizeFrenchTimestamp(raw) ?? raw;
}

function toNameSet(value: unknown): Set<string> {
  if (Array.isArray(value)) {
    return new Set(value.filter((v): v is string => typeof v === "string"));
  }
  if (typeof value === "string") {
    return new Set(
      value.split(";").map((part) => part.trim()).filter((part) => part !== ""),
    );
  }
  return new Set();
}

function coordinatesOf(record: RawStationRecord): GeoPoint | null {
  if (record.geom && typeof record.geom.lat === "number") {
    return { latitude: record.geom.lat, longitude: record.geom.lon };
  }
  const lat = numberOrNull(record.latitude);
  const lon = numberOrNull(record.longitude);
  if (lat === null || lon === null) return null;
  const scaled = Math.abs(lat) > 1000 || Math.abs(lon) > 1000;
  return scaled
    ? { latitude: lat / 1e5, longitude: lon / 1e5 }
    : { latitude: lat, longitude: lon };
}

function classifyFuel(
  fuel: FuelCode,
  record: RawStationRecord,
  available: ReadonlySet<string>,
  unavailable: ReadonlySet<string>,
): FuelState {
  const prefix = FUEL_FIELDS[fuel];
  const price = numberOrNull(record[`${prefix}_prix`]);
  const updatedAt = timestampOrNull(record[`${prefix}_maj`]);
  const ruptureType = stringOrNull(record[`${prefix}_rupture_type`]);
  const ruptureStart = timestampOrNull(record[`${prefix}_rupture_debut`]);

  let status: Availability;
  if (unavailable.has(fuel) || ruptureType !== null) {
    status = "rupture";
  } else if (available.has(fuel) || price !== null) {
    status = "available";
  } else {
    status = "not_offered";
  }

  if (status === "rupture") {
    return {
      status,
      price: null,
      updatedAt,
      ruptureType,
      ruptureStart,
    };
  }
  if (status === "available") {
    return {
      status,
      price,
      updatedAt,
      ruptureType: null,
      ruptureStart: null,
    };
  }
  return {
    status,
    price: null,
    updatedAt: null,
    ruptureType: null,
    ruptureStart: null,
  };
}

export function normalizeStation(
  record: RawStationRecord,
  origin: GeoPoint,
  fuels: readonly FuelCode[],
): StationSnapshot | null {
  const point = coordinatesOf(record);
  if (!point) return null;

  const available = toNameSet(record.carburants_disponibles);
  const unavailable = toNameSet(record.carburants_indisponibles);

  const fuelMap = new Map<FuelCode, FuelState>();
  for (const fuel of fuels) {
    fuelMap.set(
      fuel,
      classifyFuel(fuel, record, available, unavailable),
    );
  }

  const address = stringOrNull(record.adresse) ?? "";
  const city = stringOrNull(record.ville) ?? "";
  const name = [address, city].filter((part) => part !== "").join(", ") ||
    `Station ${record.id}`;

  return {
    stationId: record.id,
    name,
    address,
    city,
    latitude: point.latitude,
    longitude: point.longitude,
    distanceKm: haversineKm(origin, point),
    fuels: fuelMap,
  };
}

export function normalizeStations(
  records: readonly RawStationRecord[],
  origin: GeoPoint,
  fuels: readonly FuelCode[],
): StationSnapshot[] {
  const stations: StationSnapshot[] = [];
  for (const record of records) {
    const station = normalizeStation(record, origin, fuels);
    if (station) stations.push(station);
  }
  return stations;
}

File: src-tauri/src/ingest/normalizer.rs
use std::collections::{BTreeMap, HashSet};

use chrono::{LocalResult, NaiveDate, SecondsFormat, TimeZone, Utc};
use chrono_tz::Europe::Paris;
use regex::Regex;
use serde_json::Value;

use crate::domain::geo::{haversine_km, GeoPoint};
use crate::domain::model::{Availability, FuelCode, FuelState, StationSnapshot};

use super::raw_record::RawStationRecord;

fn fuel_prefix(fuel: FuelCode) -> &'static str {
    match fuel {
        FuelCode::Gazole => "gazole",
        FuelCode::Sp95 => "sp95",
        FuelCode::E10 => "e10",
        FuelCode::Sp98 => "sp98",
        FuelCode::E85 => "e85",
        FuelCode::Gplc => "gplc",
    }
}

fn number_or_null(value: Option<&Value>) -> Option<f64> {
    match value {
        Some(Value::Number(number)) => number.as_f64().filter(|n| n.is_finite()),
        Some(Value::String(text)) if !text.trim().is_empty() => {
            text.parse::<f64>().ok().filter(|n| n.is_finite())
        }
        _ => None,
    }
}

fn string_or_null(value: Option<&Value>) -> Option<String> {
    match value {
        Some(Value::String(text)) => {
            let trimmed = text.trim();
            if trimmed.is_empty() {
                None
            } else {
                Some(trimmed.to_string())
            }
        }
        _ => None,
    }
}

/// The ODS feed returns French station timestamps as local wall-clock time but
/// labels them UTC (e.g. "2026-09-18T18:49:25+00:00"). Reinterpret the
/// wall-clock as Europe/Paris time so the resulting instant is correct.
pub fn normalize_french_timestamp(value: &str) -> Option<String> {
    let pattern =
        Regex::new(r"^(\d{4})-(\d{2})-(\d{2})[T ](\d{2}):(\d{2})(?::(\d{2}))?").ok()?;
    let captures = pattern.captures(value)?;

    let year: i32 = captures.get(1)?.as_str().parse().ok()?;
    let month: u32 = captures.get(2)?.as_str().parse().ok()?;
    let day: u32 = captures.get(3)?.as_str().parse().ok()?;
    let hour: u32 = captures.get(4)?.as_str().parse().ok()?;
    let minute: u32 = captures.get(5)?.as_str().parse().ok()?;
    let second: u32 = captures
        .get(6)
        .and_then(|part| part.as_str().parse().ok())
        .unwrap_or(0);

    let date = NaiveDate::from_ymd_opt(year, month, day)?;
    let naive = date.and_hms_opt(hour % 24, minute, second)?;
    let localized = match Paris.from_local_datetime(&naive) {
        LocalResult::Single(datetime) => datetime,
        LocalResult::Ambiguous(datetime, _) => datetime,
        LocalResult::None => return None,
    };
    Some(
        localized
            .with_timezone(&Utc)
            .to_rfc3339_opts(SecondsFormat::Millis, true),
    )
}

fn timestamp_or_null(value: Option<&Value>) -> Option<String> {
    let raw = string_or_null(value)?;
    Some(normalize_french_timestamp(&raw).unwrap_or(raw))
}

fn to_name_set(value: Option<&Value>) -> HashSet<String> {
    match value {
        Some(Value::Array(items)) => items
            .iter()
            .filter_map(|item| item.as_str().map(|text| text.to_string()))
            .collect(),
        Some(Value::String(text)) => text
            .split(';')
            .map(|part| part.trim())
            .filter(|part| !part.is_empty())
            .map(|part| part.to_string())
            .collect(),
        _ => HashSet::new(),
    }
}

fn coordinates_of(record: &RawStationRecord) -> Option<GeoPoint> {
    if let Some(geom) = &record.geom {
        if let (Some(lat), Some(lon)) = (geom.lat, geom.lon) {
            return Some(GeoPoint {
                latitude: lat,
                longitude: lon,
            });
        }
    }
    let lat = number_or_null(record.latitude.as_ref())?;
    let lon = number_or_null(record.longitude.as_ref())?;
    let scaled = lat.abs() > 1000.0 || lon.abs() > 1000.0;
    Some(if scaled {
        GeoPoint {
            latitude: lat / 1e5,
            longitude: lon / 1e5,
        }
    } else {
        GeoPoint {
            latitude: lat,
            longitude: lon,
        }
    })
}

fn classify_fuel(
    fuel: FuelCode,
    record: &RawStationRecord,
    available: &HashSet<String>,
    unavailable: &HashSet<String>,
) -> FuelState {
    let prefix = fuel_prefix(fuel);
    let price = number_or_null(record.extra.get(&format!("{prefix}_prix")));
    let updated_at = timestamp_or_null(record.extra.get(&format!("{prefix}_maj")));
    let rupture_type = string_or_null(record.extra.get(&format!("{prefix}_rupture_type")));
    let rupture_start = timestamp_or_null(record.extra.get(&format!("{prefix}_rupture_debut")));

    let status = if unavailable.contains(fuel.as_str()) || rupture_type.is_some() {
        Availability::Rupture
    } else if available.contains(fuel.as_str()) || price.is_some() {
        Availability::Available
    } else {
        Availability::NotOffered
    };

    match status {
        Availability::Rupture => FuelState {
            status,
            price: None,
            updated_at,
            rupture_type,
            rupture_start,
        },
        Availability::Available => FuelState {
            status,
            price,
            updated_at,
            rupture_type: None,
            rupture_start: None,
        },
        Availability::NotOffered => FuelState {
            status,
            price: None,
            updated_at: None,
            rupture_type: None,
            rupture_start: None,
        },
    }
}

pub fn normalize_station(
    record: &RawStationRecord,
    origin: GeoPoint,
    fuels: &[FuelCode],
) -> Option<StationSnapshot> {
    let point = coordinates_of(record)?;

    let available = to_name_set(record.extra.get("carburants_disponibles"));
    let unavailable = to_name_set(record.extra.get("carburants_indisponibles"));

    let mut fuel_map = BTreeMap::new();
    for fuel in fuels {
        fuel_map.insert(*fuel, classify_fuel(*fuel, record, &available, &unavailable));
    }

    let address = string_or_null(record.adresse.as_ref()).unwrap_or_default();
    let city = string_or_null(record.ville.as_ref()).unwrap_or_default();
    let name = {
        let parts: Vec<&str> = [address.as_str(), city.as_str()]
            .into_iter()
            .filter(|part| !part.is_empty())
            .collect();
        if parts.is_empty() {
            format!("Station {}", record.id)
        } else {
            parts.join(", ")
        }
    };

    Some(StationSnapshot {
        station_id: record.id,
        name,
        address,
        city,
        latitude: point.latitude,
        longitude: point.longitude,
        distance_km: haversine_km(origin, point),
        fuels: fuel_map,
    })
}

pub fn normalize_stations(
    records: &[RawStationRecord],
    origin: GeoPoint,
    fuels: &[FuelCode],
) -> Vec<StationSnapshot> {
    records
        .iter()
        .filter_map(|record| normalize_station(record, origin, fuels))
        .collect()
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::json;

    fn fuels() -> Vec<FuelCode> {
        FuelCode::ALL.to_vec()
    }

    fn origin() -> GeoPoint {
        GeoPoint {
            latitude: 48.8566,
            longitude: 2.3522,
        }
    }

    fn record(value: Value) -> RawStationRecord {
        serde_json::from_value(value).unwrap()
    }

    #[test]
    fn classifies_available_rupture_and_not_offered() {
        let raw = record(json!({
            "id": 89100001,
            "geom": { "lon": 3.309, "lat": 48.183 },
            "adresse": "84 ROUTE DE MAILLOT",
            "ville": "Sens",
            "gazole_prix": 2.469,
            "gazole_maj": "2026-09-18T09:07:28+00:00",
            "gazole_rupture_type": null,
            "sp95_prix": null,
            "sp95_rupture_type": "definitive",
            "sp95_rupture_debut": "2022-11-26T09:30:22+00:00",
            "carburants_disponibles": ["Gazole", "E85", "E10", "SP98"],
            "carburants_indisponibles": ["SP95", "GPLc"]
        }));

        let station = normalize_station(&raw, origin(), &fuels()).unwrap();
        assert_eq!(station.fuels[&FuelCode::Gazole].status, Availability::Available);
        assert_eq!(station.fuels[&FuelCode::Gazole].price, Some(2.469));
        assert_eq!(station.fuels[&FuelCode::Sp95].status, Availability::Rupture);
        assert_eq!(
            station.fuels[&FuelCode::Sp95].rupture_type.as_deref(),
            Some("definitive")
        );
        assert_eq!(station.fuels[&FuelCode::Gplc].status, Availability::Rupture);
        assert_eq!(station.fuels[&FuelCode::E10].status, Availability::Available);
        assert_eq!(station.fuels[&FuelCode::E10].price, None);
    }

    #[test]
    fn treats_a_fuel_with_no_price_and_no_rupture_as_not_offered() {
        let raw = record(json!({
            "id": 1,
            "geom": { "lon": 2.35, "lat": 48.85 },
            "adresse": "A",
            "ville": "B",
            "carburants_disponibles": null,
            "carburants_indisponibles": null
        }));
        let station = normalize_station(&raw, origin(), &fuels()).unwrap();
        assert_eq!(station.fuels[&FuelCode::Sp98].status, Availability::NotOffered);
        assert_eq!(station.fuels[&FuelCode::Sp98].price, None);
    }

    #[test]
    fn falls_back_to_scaled_integer_coordinates() {
        let raw = record(json!({
            "id": 2,
            "latitude": "4818300",
            "longitude": "330900",
            "adresse": "A",
            "ville": "B"
        }));
        let station = normalize_station(&raw, origin(), &fuels()).unwrap();
        assert_eq!(station.latitude, 48.183);
        assert_eq!(station.longitude, 3.309);
    }

    #[test]
    fn computes_a_plausible_distance_from_the_origin() {
        let raw = record(json!({ "id": 3, "geom": { "lon": 2.3522, "lat": 48.8566 } }));
        let station = normalize_station(&raw, origin(), &fuels()).unwrap();
        assert_eq!(station.distance_km, 0.0);
    }

    #[test]
    fn parses_semicolon_delimited_fuel_lists() {
        let raw = record(json!({
            "id": 4,
            "geom": { "lon": 2.35, "lat": 48.85 },
            "carburants_disponibles": "Gazole;E10",
            "carburants_indisponibles": "SP95",
            "sp95_rupture_type": "temporaire"
        }));
        let station = normalize_station(&raw, origin(), &fuels()).unwrap();
        assert_eq!(station.fuels[&FuelCode::Gazole].status, Availability::Available);
        assert_eq!(station.fuels[&FuelCode::E10].status, Availability::Available);
        assert_eq!(station.fuels[&FuelCode::Sp95].status, Availability::Rupture);
        assert_eq!(
            station.fuels[&FuelCode::Sp95].rupture_type.as_deref(),
            Some("temporaire")
        );
    }

    #[test]
    fn reinterprets_french_local_timestamps_labelled_as_utc() {
        assert_eq!(
            normalize_french_timestamp("2026-09-18T18:49:25+00:00").as_deref(),
            Some("2026-09-18T16:49:25.000Z")
        );
        assert_eq!(
            normalize_french_timestamp("2026-01-15T08:30:00+00:00").as_deref(),
            Some("2026-01-15T07:30:00.000Z")
        );
        assert_eq!(
            normalize_french_timestamp("2026-09-18 18:49:25").as_deref(),
            Some("2026-09-18T16:49:25.000Z")
        );
        assert_eq!(normalize_french_timestamp("not a date"), None);
    }
}
